segpy/field_types.py

```diff
@@ -1,29 +1,86 @@
-class Int16(int):
-
-    MINIMUM = -32768
-    MAXIMUM = 32767
-    SIZE = 2
-    SEG_Y_TYPE = 'int16'
-
-    def __new__(cls, *args, **kwargs):
-        instance = super().__new__(cls, *args, **kwargs)
-        if not (Int16.MINIMUM <= instance <= Int16.MAXIMUM):
-            raise ValueError("{} value {!r} outside range {} to {}".format(cls.__name__, instance,
-                                                                        cls.MINIMUM, cls.MAXIMUM))
+from .datatypes import LIMITS, SEG_Y_TYPE_TO_CTYPE, size_in_bytes
+
+
+class IntFieldMeta(type):
+    """Metaclass for signed and unsigned int fields.
+    """
+
+    def class_new(cls, *args, **kwargs):
+        instance = super(cls, cls).__new__(cls, *args, **kwargs)
+        if not (cls.MINIMUM <= instance <= cls.MAXIMUM):
+            raise ValueError("{} value {!r} outside range {} to {}".format(
+                cls.__name__, instance,
+                cls.MINIMUM, cls.MAXIMUM))
         return instance
 
+    def __new__(cls, name, bases, namespace,
+                seg_y_type,
+                min_value=None,
+                max_value=None,
+                **kwargs):
+
+        bases = bases + (int,)
+
+        if min_value is None:
+            min_value = LIMITS[seg_y_type].min
+
+        if max_value is None:
+            max_value = LIMITS[seg_y_type].max
+
+        namespace.update({
+            'SEG_Y_TYPE': seg_y_type,
+            'MINIMUM': min_value,
+            'MAXIMUM': max_value,
+            'SIZE': size_in_bytes(SEG_Y_TYPE_TO_CTYPE[seg_y_type]),
+            '__new__': cls.class_new,
+        })
+        return super().__new__(cls, name, bases, namespace, **kwargs)
+
+    def __init__(cls, name, bases, namespace, *args, **kwargs):
+        super().__init__(name, bases, namespace)
+
+
+class Int16(metaclass=IntFieldMeta,
+            seg_y_type='int16'):
+    pass
 
-class Int32(int):
 
-    MINIMUM = -2147483648
-    MAXIMUM = 2147483647
-    SIZE = 4
-    SEG_Y_TYPE = 'int32'
+class UInt16(metaclass=IntFieldMeta,
+             seg_y_type='uint16'):
+    pass
 
-    def __new__(cls, *args, **kwargs):
-        instance = super().__new__(cls, *args, **kwargs)
-        if not (Int32.MINIMUM <= instance <= Int32.MAXIMUM):
-            raise ValueError("{} value {!r} outside range {} to {}".format(cls.__name__, instance,
-                                                                        cls.MINIMUM, cls.MAXIMUM))
+
+class Int32(metaclass=IntFieldMeta,
+            seg_y_type='int32'):
+    pass
+
+
+class UInt32(metaclass=IntFieldMeta,
```

Comment:
I know I suggested this, but now I'm wondering if `UInt32` is the right name given that it's really non-negative `Int32`. `PInt32` for "Positive-int' (although we admit zero too...this is probably clear enough?)

Reply:
What about UInt31?

Reply:
NNInt32 - Non-Negative Int32. Horrible but accurate.